File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        int[] dy = {0 , -1, 0 , 1};
        int[] dx = {1, 0 , -1 , 0};
        LinkedList<int[]> que = new LinkedList();
        int[] yx = {0,0,0};
        que.addLast(yx);
        while(!que.isEmpty()){
            int[] dot = que.pollFirst();
            if (dot[0] == maps.length - 1 && dot[1] == maps[0].length - 1){
                return dot[2] + 1;
            }
            
            for(int i = 0; i < 4; i++){
                int ny = dot[0] + dy[i];
                int nx = dot[1] + dx[i];
                if(0<= ny && ny < maps.length &&
                   0 <= nx && nx < maps[0].length &&
                   maps[ny][nx] == 1 &&
                   !visited[ny][nx]){
                    visited[ny][nx] = true;
                    que.addLast(new int[] {ny,nx, dot[2] + 1});
                }
                    
                
            }
        }
        
        
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque


class Length:
    maps = []
    visited = []
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]

    def __init__(self, maps):
        self.maps = maps
        self.visited = [[False] * len(c) for c in maps]

    def isInnerPoint(self, y, x):
        return (
            0 <= y < len(self.maps)
            and 0 <= x < len(self.maps[0])
            and not self.visited[y][x]
            and self.maps[y][x] == 1
        )

    def isTarget(self, y, x):
        return y == len(self.maps) - 1 and x == len(self.maps[0]) - 1

    def bfs(self):
        que = deque([])
        que.append((0, 0, 1))
        while que:
            y, x, cnt = que.popleft()
            if self.isTarget(y, x):
                return cnt

            for i in range(4):
                ny = y + self.dy[i]
                nx = x + self.dx[i]
                if self.isInnerPoint(ny, nx):
                    self.visited[ny][nx] = True
                    que.append((ny, nx, cnt + 1))

        return -1


def solution(maps):
    return Length(maps).bfs()
